src/modules/image-upload/image-upload.js

```diff
@@ -69,41 +103,112 @@ function getSignature(policy) {
+		upload_url = `https://${config.s3.uploadBucket}/`,
+		thumbnail_url = `https://${config.s3.thumbnailsBucket}/`,
+		filename = policyReq.filename;
+	let key = keyPrefix,
+		thumbnail = thumbnail_url + keyPrefix,
+		url;
+
+	if (/\s{2,}/.test(filename)) {
+		throw new Error('S3 does not support filenames with multiple spaces');
+	}
+
+	const ext = filename.split('.').pop() || 'jpg';
+
+	switch (policyReq.uploadType) {
+	case 'avatar':
+		key += 'avatar.' + ext;
+		url = upload_url + key;
+		thumbnail += '256.jpg';
+		break;
+	case 'content':
+		key += 'content.' + ext;
+		url = upload_url + key;
+		thumbnail += '320.jpg';
+		break;
+	}
 
-	policyReq.response = {
-		acl: config.s3.acl,
+	return {
+		request_url: upload_url,
+		original: url,
+		thumbnail,
+		policy: {
+			key,
+			acl: config.s3.acl,
+			policy,
+			success_action_status: '201',
+			'x-amz-algorithm': config.s3.algorithm,
+			'x-amz-credential': getCredential(),
+			'x-amz-date': getDate({ long: true }),
+			'x-amz-signature': signature,
+		},
+	};
+}
+
+// For old bucket
+export function getOldResponse(policyReq) {
+	const keyPrefix = getOldKeyPrefix(policyReq.auth.user, policyReq.uploadType, policyReq.textId),
+		policy = getOldPolicy(keyPrefix),
+		signature = getOldSignature(policy);
+
+	return {
+		acl: config.oldS3.acl,
 		policy,
 		keyPrefix,
-		bucket: config.s3.bucket,
-		'x-amz-algorithm': config.s3.algorithm,
-		'x-amz-credential': getCredential(),
-		'x-amz-date': getDate(true),
+		bucket: config.oldS3.bucket,
+		'x-amz-algorithm': config.oldS3.algorithm,
+		'x-amz-credential': getOldCredential(),
+		'x-amz-date': getDate({ long: true }),
 		'x-amz-signature': signature,
 	};
 }
 
 if (!config.s3) {
 	winston.info('Image upload is disabled');
-	bus.on('s3/getPolicy', (policyReq, next) => {
+	bus.on('s3/getPolicy', (policyReq) => {
 		policyReq.response = {};
-		policyReq.response.error = new EnhancedError('Image upload is temporarily disabled', 'NO_CONFIG_FOUND_FO_S3');
-		next();
+		policyReq.response.error = new EnhancedError('Image upload is temporarily disabled', 'NO_CONFIG_FOUND_FOR_S3');
 	}, APP_PRIORITIES.IMAGE_UPLOAD);
 } else {
-	bus.on('s3/getPolicy', (policyReq, next) => {
-		getResponse(policyReq);
-		next();
+	bus.on('s3/getPolicy', (policyReq) => {
+		/*
+	     * policy requests from new clients are going to have
+	     * the filename property. use it to distinguish
+	     * the clients and handle the fallback.
+		 */
+		try {
+			if (policyReq.filename) {
+				// generate response for the new client
+				policyReq.response = getResponse(policyReq);
+			} else {
+				// generate response for the old client
+				policyReq.response = getOldResponse(policyReq);
+			}
+		} catch (error) {
+			policyReq.response = { error };
+		}
 	}, APP_PRIORITIES.IMAGE_UPLOAD);
 	winston.info('Image upload is ready');
 }
 
 const uploadImage = async (userName: string, imageUrl: string, propName: string) => {
-	const imageName = 'avatar';
```

Comment:
I think uploads from the client are 'avatar' (without suffix). Unsure if it makes a difference, though.